2019-02-28-Monad-structure/MonadEval.hs

```diff
@@ -0,0 +1,100 @@
+module MonadEval
+  where
+
+import Control.Monad.Writer
+-- |In Haskell, types are defined with the @data@ keyword, product
+-- types can be written as tuples. All types and constructors in
+-- Haskell have to start with upper case letters.
+data Term = Cons Int | Div (Term,Term)
+
+-- |Simple evaluation
+evalSim :: Term -> Int
+evalSim (Cons x) = x
+evalSim (Div (x,y)) = div (evalSim x) (evalSim y)
+
+-- |To handle division by zero and issue a message, the @Either@ type
+-- functor is used to construct the return type of the evaluation function
+-- It is defined as
+-- @data Either a b = Left a|Right b
+-- The data functor @Either a@ is a monad in Haskell, so the
+-- binding operator >>= is already defined. Note that in Haskell nameless
+-- functions are introduced with the @\args -> body@ notation.
+-- In @evalSafe@, the dangerous division branch of @evalSim@ is reimplemented
+-- and all other results are simply packed into the @Either String@ monad
+-- with @return@, which is Haskell's name for the monadic unit.
+-- Note that
+-- without monadic binding @>>=@, the @case@ match would descend into a nested
+-- hell because the results of @evalSafe@ would have to be checked for leftness
+-- and rightness in addition to zeroness. If a @Left "Division by zero"@ occurs
+-- anywhere in the monadic binding chain, that will be its final result.
+evalSafe :: Term -> Either String Int
+evalSafe (Div (x,y)) = evalSafe y >>= \ev -> case ev of
+                         0 -> Left "Division by zero"
+                         rslt -> evalSafe x >>= \nm -> return $ div nm rslt
+evalSafe (Cons x) = return x
+
+-- |For logging, the monad to use is @Writer w@ because the logging
+-- value of type @w@ is accumulated automatically when chaining through
+-- monadic values. That's why @w@ is required to be an instance of @Monoid@,
+-- which allows the accumulation of values with the monoidal composition.
+-- Whilst @State@ could be used as well, the accumulation would have to be
+-- done manually.
+-- For counting divisions, a @Writer (Sum Int) Int@ monad is
+-- called for because @Sum Int@ is the monoid with @0@ as neutral element
+-- and @+@ as operation.
+logDivs :: (Term -> Int) -> Term -> Writer (Sum Int) Int
+logDivs f arg@(Div (x,y)) = logDivs f x >>= logSecond
+  where
+    logSecond n = logDivs f y >>= combineWith n
+    combineWith nm den = writer (f $  Div (Cons nm, Cons den),1)
```

Comment:
It's an interesting idea to have `logDivs` wrap around the `evalSim`, but it looks like this implementation will apply `f` every time it recurses, leading to unnecessary evaluations. Was this intended?